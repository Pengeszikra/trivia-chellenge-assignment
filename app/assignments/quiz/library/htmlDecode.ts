import {decode} from 'html-entities';

export const htmlDecode = (input:string):string => decode(input);
