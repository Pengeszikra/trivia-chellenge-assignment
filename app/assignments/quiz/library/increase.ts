export const increase:(n:number) => number = (n:number) => n + 1;