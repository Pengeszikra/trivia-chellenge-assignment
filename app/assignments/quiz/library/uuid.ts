export const uuid:() => string
 = () => Math.random().toString(32).slice(-8)
;