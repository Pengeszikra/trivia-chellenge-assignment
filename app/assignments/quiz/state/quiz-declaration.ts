export enum PROGRESS {
  INTRO,
  QUIZ,
  RESULTS,
}

export interface IQuestionSource {
  category: string;
  type: string;
  difficulty: string;
  question: string;
  correct_answer: string;
  incorrect_answers: string[];
}

export type TAnswerId = string;

export interface IAnswer {
  answerId: TAnswerId;
  answer: string;
}

export interface IQuestion {
  questionId: string;
  question: string;
  category: string;
  shuffledAnswerList: IAnswer[];
}

export interface IQuizState {
  sourceList: IQuestionSource[];
  correctAnswerList: TAnswerId[];
  solutionList: TAnswerId[];
  shuffledQuestionList: IQuestion[];
  answerIndex: number;
  progress: PROGRESS;
}