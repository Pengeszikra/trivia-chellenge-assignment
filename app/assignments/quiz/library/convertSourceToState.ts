import { IAnswer, IQuestion, IQuestionSource, IQuizState, TAnswerId } from "../state/quiz-declaration";
import { htmlDecode } from "./htmlDecode";
import { shuffle } from "./shuffle";

export type TConvertSourceToState = (source:IQuestionSource[], idList:string[]) => Partial<IQuizState>;
export type TMakeAnswer = (answerId:string, rawAnswer:string) => IAnswer;

export const answerListShuffle:(IQuestion) => IQuestion 
  = ({shuffledAnswerList, ...rest}) => ({shuffledAnswerList: shuffle(shuffledAnswerList), ...rest});

export const convertSourceToState:TConvertSourceToState = (source = [], idList) => {

  const makeAnswer:TMakeAnswer = (answerId, rawAnswer) => ({
    answerId,
    answer: htmlDecode(rawAnswer)
  });

  const questionList:IQuestion[] =
    source
      .map(({category, question, correct_answer, incorrect_answers}) => 
        ({
          questionId: idList.pop(),
          category: htmlDecode(category),
          question: htmlDecode(question),
          shuffledAnswerList: [
              makeAnswer(idList.pop(), correct_answer),
              ...incorrect_answers.map((answer:string) => makeAnswer(idList.pop(), answer))
            ]
        }))
    ;

    const correctAnswerList:TAnswerId[] = questionList
      .map(
        ({shuffledAnswerList}) => shuffledAnswerList?.[0]?.answerId
      )
    ;

  return ({
    sourceList: source,
    correctAnswerList,
    shuffledQuestionList: shuffle(questionList.map(answerListShuffle))
  });
}
;
