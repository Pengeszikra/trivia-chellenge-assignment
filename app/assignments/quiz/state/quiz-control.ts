import { actionFactory, kebabToCamelCase } from "react-troll";
import { Action } from "../../../utils/react-troll-declaration";
import { convertSourceToState } from "../library/convertSourceToState";
import { IQuizState, PROGRESS } from "./quiz-declaration";

export const [quizActionsSet, action] = actionFactory(kebabToCamelCase);

export const
  SET_PROGRESS = action('set-progress'),
  READ_SOURCE = action('read-source'),
  PREPARE_QUIZ = action('prepare-quiz'),
  BEGIN_QUIZ = action('begin-quiz'),
  NEXT_QUIZ = action('next-quiz'),
  MARK_ANSWER = action('mark-answer'),
  PLAY_AGAIN = action('play-again')
;

export const quizInitialState:IQuizState = {
  sourceList: [],
  correctAnswerList: [],
  solutionList: [],
  shuffledQuestionList: [],
  answerIndex: 0,
  progress: PROGRESS.INTRO,
}

export type TQuziReducer = (state:IQuizState, action:Action ) => IQuizState;

export const quizReducer:TQuziReducer = (state:IQuizState, {type, payload}:Action) => {
  switch (type) {
    case SET_PROGRESS: return {...state, progress:payload};
    case READ_SOURCE: return {...state, sourceList:payload};
    case PREPARE_QUIZ: {
      if (!Array.isArray(payload)) return state;
      const idList:string[] = payload;
      return {...state, ...convertSourceToState(state.sourceList, idList)};
    }
    case BEGIN_QUIZ: return {...state, answerIndex: 0, progress:PROGRESS.QUIZ};
    case NEXT_QUIZ: {
      const answerIndex = state.answerIndex + 1;
      return {
        ...state, 
        answerIndex, 
        progress: answerIndex < state.sourceList.length  
          ? state.progress 
          : PROGRESS.RESULTS
      };
    }
    case MARK_ANSWER: return {...state, solutionList: [...state.solutionList, payload]};
    case PLAY_AGAIN: return quizInitialState;
    default: return state;
  }
}