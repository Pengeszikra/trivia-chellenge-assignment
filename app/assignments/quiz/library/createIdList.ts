import {IQuestionSource} from '../state/quiz-declaration';

export type TCreateIdList = (idCreator:()=>string) => (source:IQuestionSource[]) => string[];

export const createIdList: TCreateIdList = idCreator => source => 
  Array(source.length * 5)
    .fill('')
    .map(() => idCreator())
;