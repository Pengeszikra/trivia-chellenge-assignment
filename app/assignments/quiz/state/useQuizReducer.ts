import { useTroll } from 'react-troll';
import { quizActionsSet, quizInitialState, quizReducer } from './quiz-control';
import { IQuizState } from './quiz-declaration';
import { TrollFactory } from '../../utils/react-troll-declaration';

export const useQuizReducer:TrollFactory<IQuizState> = () => useTroll(quizReducer, quizInitialState, quizActionsSet);