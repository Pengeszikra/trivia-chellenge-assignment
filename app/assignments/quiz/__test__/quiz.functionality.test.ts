import { convertSourceToState } from "../library/convertSourceToState";
import { createIdList } from "../library/createIdList";
import { IQuizState, IQuestionSource } from "../state/quiz-declaration";

const source = [
  {
    "category": "Entertainment: Video Games",
    "type": "boolean",
    "difficulty": "hard",
    "question": "The retail disc of Tony Hawk&#039;s Pro Skater 5 only comes with the tutorial.",
    "correct_answer": "True",
    "incorrect_answers": ["False"]
  },
  {
    "category": "History",
    "type": "boolean",
    "difficulty": "hard",
    "question": "Japan was part of the Allied Powers during World War I.",
    "correct_answer": "True",
    "incorrect_answers": ["False"]
  },
  {
    "category": "Entertainment: Comics",
    "type": "boolean",
    "difficulty": "hard",
    "question": "In the comic book &quot;Archie&quot;, Betty is friends with Veronica because she is rich.",
    "correct_answer": "False",
    "incorrect_answers": ["True"]
  },
  {
    "category": "Entertainment: Video Games",
    "type": "boolean",
    "difficulty": "hard",
    "question": "In Undertale, having a &quot;Fun Value&quot; set to 56-57 will play the &quot;Wrong Number Song Call&quot;.",
    "correct_answer": "False",
    "incorrect_answers": ["True"]
  },
  {
    "category": "Politics",
    "type": "boolean",
    "difficulty": "hard",
    "question": "Joko Widodo has appeared in the cover of a TIME magazine.",
    "correct_answer": "True",
    "incorrect_answers": ["False"]
  },
  {
    "category": "Entertainment: Books",
    "type": "boolean",
    "difficulty": "hard",
    "question": "Harry Potter was born on July 31st, 1980.",
    "correct_answer": "True",
    "incorrect_answers": ["False"]
  },
  {
    "category": "Politics",
    "type": "boolean",
    "difficulty": "hard",
    "question": "Nazi Germany surrendered on Harry Truman&#039;s birthday while he was president.",
    "correct_answer": "True",
    "incorrect_answers": [
        "False"
    ]
  },
  {
    "category": "History",
    "type": "boolean",
    "difficulty": "hard",
    "question": "During the Winter War, the amount of Soviet Union soliders that died or went missing was five times more than Finland&#039;s.",
    "correct_answer": "True",
    "incorrect_answers": [
        "False"
    ]
  },
  {
    "category": "Entertainment: Film",
    "type": "boolean",
    "difficulty": "hard",
    "question": "The weapon Clint Eastwood uses in &quot;Dirty Harry&quot; was a .44 Automag.",
    "correct_answer": "False",
    "incorrect_answers": ["True"]
  },
  {
    "category": "Entertainment: Japanese Anime & Manga",
    "type": "boolean",
    "difficulty": "hard",
    "question": "The character Plum from &quot;No Game No Life&quot; is a girl.",
    "correct_answer": "False",
    "incorrect_answers": ["True"]
  }
];

test ('simple addition', () => {
  expect (
    source.at(-1).question
  ).toStrictEqual(
    "The character Plum from &quot;No Game No Life&quot; is a girl."
  );
});

test ('create enough id for source', () => {
  let firstId = 100;
  const mockIdCreator = () => (firstId++).toString(); 
  
  expect (
    createIdList(mockIdCreator)(source.slice(0, 2))
    
  ).toStrictEqual(
    ['100', '101', '102', '103', '104', '105', '106', '107', '108', '109']
  );
});

test ('convert IQuestionSourceList to IQuizState', () => {
  let firstId = 100;
  const mockIdCreator = () => (firstId++).toString(); 
  const idList = createIdList(mockIdCreator)(source);

  expect (
    convertSourceToState(source, idList)
  ).toMatchSnapshot();
});
