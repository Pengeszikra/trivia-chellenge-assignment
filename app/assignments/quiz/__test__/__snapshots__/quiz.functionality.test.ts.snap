// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert IQuestionSourceList to IQuizState 1`] = `
Object {
  "correctAnswerList": Array [
    "381",
    "378",
    "375",
    "372",
    "369",
    "366",
    "363",
    "360",
    "357",
    "354",
  ],
  "shuffledQuestionList": Array [
    Object {
      "category": "Politics",
      "question": "Nazi Germany surrendered on Harry Truman's birthday while he was president.",
      "questionId": "364",
      "shuffledAnswerList": Array [
        Object {
          "answer": "True",
          "answerId": "363",
        },
        Object {
          "answer": "False",
          "answerId": "362",
        },
      ],
    },
    Object {
      "category": "Entertainment: Video Games",
      "question": "The retail disc of Tony Hawk's Pro Skater 5 only comes with the tutorial.",
      "questionId": "382",
      "shuffledAnswerList": Array [
        Object {
          "answer": "True",
          "answerId": "381",
        },
        Object {
          "answer": "False",
          "answerId": "380",
        },
      ],
    },
    Object {
      "category": "Entertainment: Comics",
      "question": "In the comic book \\"Archie\\", Betty is friends with Veronica because she is rich.",
      "questionId": "376",
      "shuffledAnswerList": Array [
        Object {
          "answer": "False",
          "answerId": "375",
        },
        Object {
          "answer": "True",
          "answerId": "374",
        },
      ],
    },
    Object {
      "category": "Entertainment: Books",
      "question": "Harry Potter was born on July 31st, 1980.",
      "questionId": "367",
      "shuffledAnswerList": Array [
        Object {
          "answer": "False",
          "answerId": "365",
        },
        Object {
          "answer": "True",
          "answerId": "366",
        },
      ],
    },
    Object {
      "category": "Entertainment: Film",
      "question": "The weapon Clint Eastwood uses in \\"Dirty Harry\\" was a .44 Automag.",
      "questionId": "358",
      "shuffledAnswerList": Array [
        Object {
          "answer": "True",
          "answerId": "356",
        },
        Object {
          "answer": "False",
          "answerId": "357",
        },
      ],
    },
    Object {
      "category": "History",
      "question": "Japan was part of the Allied Powers during World War I.",
      "questionId": "379",
      "shuffledAnswerList": Array [
        Object {
          "answer": "False",
          "answerId": "377",
        },
        Object {
          "answer": "True",
          "answerId": "378",
        },
      ],
    },
    Object {
      "category": "Entertainment: Video Games",
      "question": "In Undertale, having a \\"Fun Value\\" set to 56-57 will play the \\"Wrong Number Song Call\\".",
      "questionId": "373",
      "shuffledAnswerList": Array [
        Object {
          "answer": "True",
          "answerId": "371",
        },
        Object {
          "answer": "False",
          "answerId": "372",
        },
      ],
    },
    Object {
      "category": "Entertainment: Japanese Anime & Manga",
      "question": "The character Plum from \\"No Game No Life\\" is a girl.",
      "questionId": "355",
      "shuffledAnswerList": Array [
        Object {
          "answer": "True",
          "answerId": "353",
        },
        Object {
          "answer": "False",
          "answerId": "354",
        },
      ],
    },
    Object {
      "category": "History",
      "question": "During the Winter War, the amount of Soviet Union soliders that died or went missing was five times more than Finland's.",
      "questionId": "361",
      "shuffledAnswerList": Array [
        Object {
          "answer": "False",
          "answerId": "359",
        },
        Object {
          "answer": "True",
          "answerId": "360",
        },
      ],
    },
    Object {
      "category": "Politics",
      "question": "Joko Widodo has appeared in the cover of a TIME magazine.",
      "questionId": "370",
      "shuffledAnswerList": Array [
        Object {
          "answer": "True",
          "answerId": "369",
        },
        Object {
          "answer": "False",
          "answerId": "368",
        },
      ],
    },
  ],
  "sourceList": Array [
    Object {
      "category": "Entertainment: Video Games",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "The retail disc of Tony Hawk&#039;s Pro Skater 5 only comes with the tutorial.",
      "type": "boolean",
    },
    Object {
      "category": "History",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "Japan was part of the Allied Powers during World War I.",
      "type": "boolean",
    },
    Object {
      "category": "Entertainment: Comics",
      "correct_answer": "False",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "True",
      ],
      "question": "In the comic book &quot;Archie&quot;, Betty is friends with Veronica because she is rich.",
      "type": "boolean",
    },
    Object {
      "category": "Entertainment: Video Games",
      "correct_answer": "False",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "True",
      ],
      "question": "In Undertale, having a &quot;Fun Value&quot; set to 56-57 will play the &quot;Wrong Number Song Call&quot;.",
      "type": "boolean",
    },
    Object {
      "category": "Politics",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "Joko Widodo has appeared in the cover of a TIME magazine.",
      "type": "boolean",
    },
    Object {
      "category": "Entertainment: Books",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "Harry Potter was born on July 31st, 1980.",
      "type": "boolean",
    },
    Object {
      "category": "Politics",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "Nazi Germany surrendered on Harry Truman&#039;s birthday while he was president.",
      "type": "boolean",
    },
    Object {
      "category": "History",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "During the Winter War, the amount of Soviet Union soliders that died or went missing was five times more than Finland&#039;s.",
      "type": "boolean",
    },
    Object {
      "category": "Entertainment: Film",
      "correct_answer": "False",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "True",
      ],
      "question": "The weapon Clint Eastwood uses in &quot;Dirty Harry&quot; was a .44 Automag.",
      "type": "boolean",
    },
    Object {
      "category": "Entertainment: Japanese Anime & Manga",
      "correct_answer": "False",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "True",
      ],
      "question": "The character Plum from &quot;No Game No Life&quot; is a girl.",
      "type": "boolean",
    },
  ],
}
`;
